tracks/python/exercises/book-store/mentoring.md

```diff
@@ -0,0 +1,301 @@
+items:
+
+```python
+from itertools import combinations
+
+DISCOUNTS = {
+    1:  0.0,
+    2:  0.05,
+    3:  0.1,
+    4:  0.2,
+    5:  0.25,
+}
+TITLES = (1, 2, 3, 4, 5)
+FULL_PRICE = 800 # cents
+
+def total(basket):
+    """Find the best price attainable for the given list of book titles."""
+    return best_price_for(count_quantities(basket))
+
+def best_price_for(quantities):
+    """Find the best price for the given quantities of titles."""
+    if sum(quantities) == 0:
+        return 0
+    return min(
+        combo_price(combo) + best_price_for(remove(quantities, combo))
+        for combo_size in DISCOUNTS
+        for combo in combinations(TITLES, combo_size)
+        if have_enough_for(quantities, combo)
+    )
+```
+
+Note use of `itertools.combinations`. The itertools module is very useful; encourage the mentee to
+have a look at the documentation.
+
+
+This implementation uses several helper functions:
+
+```python
+def count_quantities(book_list):
+    return tuple(book_list.count(title) for title in TITLES)
+
+def have_enough_for(quantities, items):
+    return all(
+        quantity >= items.count(title)
+        for (title, quantity) in zip(TITLES, quantities)
+    )
+
+def combo_price(combo):
+    num_books = len(combo)
+    num_titles = len(set(combo))
+    assert num_books == num_titles, "combo titles must be unique"
+    price_per_book = FULL_PRICE * (1 - DISCOUNTS[len(combo)])
+    return num_books * price_per_book
+
+def remove(quantities, items):
+    return tuple(
+        quantity - items.count(title)
+        for (title, quantity) in zip(TITLES, quantities)
+    )
+```
+
+
+Why break it up like this?
+
+-   Extensibility - e.g. it would be easy to add different prices by changing the implementation
+    of `combo_price`;
+
+-   Code reuse - the same helper functions are used in the iterative solutions below;
+
+-   Testing - we could write unit tests for each of the helper functions;
+
+-   Readability - the encapsulation provided by the helper functions makes the main function more
+    declarative;
+
+-   Functional programming - less mutability make the code less taxing to read (i.e. it doesn't
+    use as much of your working memory as you don't need to mentally simulate the evolution of the
+    state of the program).
+
+
+What about performance?
+
+-   Currently, this implementation is extremely slow. But, here's an easy optimization: use a cache.
+
+-   Why does this produce such a significant improvement in performance? There are different ways to
+    end up at the same point in the search space; it turns tree-search into graph-search. This is
+    the essence of dynamic programming.
+
+-   The runtime of the uncached function grows exponentially w.r.t.  `sum(quantities)`; but the
+    cached function's cache certainly grows no larger than `product(quantities)`.
+
+
+You could implement the cache yourself:
+
+```python
+def best_price_for(quantities, _cache=dict()):
```

Comment:
I'd much rather see the cache implemented as a global then show people the mutable default value trick. You have the note afterwards to point out the issue, but to say it "can" be a source of bugs is an understatement ... the only "advantage" of the mutable default is that it's harder for the user to purge the cache.